_posts/2020-08-01-data-science-meets-devops.md

```diff
@@ -0,0 +1,769 @@
+
+
+
+1. The [first lambda](https://github.com/kubeflow/code-intelligence/blob/faeb65757214ac93259f417b81e9e2fedafaebda/Label_Microservice/go/cmd/automl/pkg/server/server.go#L109) determines whether we need to retrain based on the age of the most recent model
+2. The [second lambda](https://github.com/kubeflow/code-intelligence/blob/faeb65757214ac93259f417b81e9e2fedafaebda/Label_Microservice/go/cmd/automl/pkg/server/server.go#L49) determines whether the model needs to be updated by comparing the most recently trained model to the model listed in a config map checked into source control.
+
+We wrap these lambdas in a simple web server and deploy on Kubernetes. One reason we chose this approach is because we wanted to rely on Kubernetes’ [git-sync](https://github.com/kubernetes/git-sync) to mirror our repository to a pod volume. This makes our lambdas super simple because all the git management is taken care of by a side-car running [git-sync](https://github.com/kubernetes/git-sync).
+
+
+#### Actuation
+
+To apply the changes necessary, we use Tekton to glue together various CLIs that we use to perform the various steps.
+
+To train our model we have a [Tekton task ](https://github.com/kubeflow/code-intelligence/blob/faeb65757214ac93259f417b81e9e2fedafaebda/tekton/tasks/run-notebook-task.yaml#L34)that
+
+
+
+1. Runs our notebook using [papermill](https://github.com/nteract/papermill).
+2. Converts the notebook to html using [nbconvert](https://nbconvert.readthedocs.io/en/latest/).
+3. Uploads the ipynb and html files to GCS using [gsutil](https://cloud.google.com/storage/docs/gsutil)
+
+To deploy our model we have a [Tekton task](https://github.com/kubeflow/code-intelligence/blob/faeb65757214ac93259f417b81e9e2fedafaebda/tekton/tasks/update-model-pr-task.yaml#L68) that
+
+
+
+1. Uses kpt to update our configmap with the desired value
+2. Runs git to push our changes to a branch
+3. Uses a wrapper around the [GitHub CLI](https://github.com/cli/cli) (gh) to create a PR
+
+We picked Tekton because the primary challenge we faced was sequentially running a series of CLIs in various containers. Tekton is perfect for this. Importantly, all the steps in a Tekton task run on the same pod which allows data to be shared between steps using a pod volume. 
+
+
+#### The Control Loop
+
+Finally, we needed to build a control loop that would periodically invoke our lambdas and launch our Tekton pipelines as needed. We used kubebuilder to create a [simple custom controller](https://github.com/kubeflow/code-intelligence/tree/master/Label_Microservice/go). Our controller’s reconcile loop will call our lambda to determines whether a sync is needed and if so with what parameters. If a sync is needed the controller fires off a Tekton pipeline to perform the actual update. An example of our custom resource is illustrated below
+
+
+```
+apiVersion: automl.cloudai.kubeflow.org/v1alpha1
```

Comment:
I'd add a comment that this APIVersion is very odd!

Reply:
Ha. Done.